2024/BIN-2024-0006.md

```diff
@@ -0,0 +1,156 @@
+| BIN-2024-006  | `OP_PAIRCOMMIT`
+| :------------ | :-------
+| Revision      | 005 (2024-10-31)
+| Author        | moonsettler `<moonsettler@protonmail.com>`
+| |
+| Layer         | Consensus (soft fork)
+| Status        | Draft
+| License       | BSD-3-CLAUSE
+| |
+| Discussion    | [https://delvingbitcoin.org/t/op-paircommit-as-a-candidate-for-addition-to-lnhance/1216/12]
+| Aliases       | [BIPs PR#????(https://github.com/bitcoin/bips/pull/?)]
+
+## Abstract
+
+This BIP describes a new tapscript opcode `OP_PAIRCOMMIT` which
+provide limited vector committment functionality.
+
+When evaluated, the `OP_PAIRCOMMIT` instruction:
+* Pops the top two values off the stack,
+* takes the "PairCommit" tagged SHA256 hash of the stack elements,
+* pushes the resulting committment on the top of the stack.
+
+## Specification
+
+`OP_PAIRCOMMIT` pops two elements off the stack, then concatenates them along with their size committments and takes the tagged SHA256 hash of that concatenated string, then pushes the resulting hash back on the stack.
+
+Given the stack `[x1, x2]`, where `x2` is at the top of the stack:
+
+`OP_PAIRCOMMIT` will push `SHA256(tagPC|x1|x2|len(x1)|pad|len(x2)|pad)` onto the stack.
```

Comment:
I was thinking a lot about how the 7,32 and 32,32 byte stack elements would get hashed, and originally wanted to decrease the number of completely zero 32 bit parts, also wanted to increase the number of bits changing for length extension attacks, somewhere along the line the 64 bit sizes would have their most significant bits set to 1, and this is giving the same preimage but cleaner and slightly more optimized.

After further thinking about this, I had to conclude that it's not necessary even if only 1 bit changes in the blocks SHA256 should still be secure and collision resistant.

So it can be simplified by removing the padding that just breaks the test vectors, I would have to redo them.